addons/marketing_card/models/mailing_mailing.py

```diff
@@ -0,0 +1,72 @@
+    card_requires_sync_count = fields.Integer(compute="_compute_card_requires_sync_count")
+    card_campaign_id = fields.Many2one('card.campaign')
+
+    @api.constrains('card_campaign_id', 'mailing_domain', 'mailing_model_id')
+    def _check_mailing_domain(self):
+        for mailing in self:
+            if mailing.card_campaign_id:
+                if mailing.sudo().mailing_model_id.model != mailing.card_campaign_id.res_model:
+                    raise exceptions.ValidationError(_(
+                        "Card Campaign Mailing should target model %(model_name)s",
+                        model_name=self.env['ir.model']._get(mailing.card_campaign_id.res_model).display_name
+                    ))
+
+    @api.depends('card_campaign_id')
+    def _compute_mailing_model_id(self):
+        for mailing in self.filtered('card_campaign_id'):
+            mailing.mailing_model_id = self.env['ir.model']._get_id(mailing.card_campaign_id.res_model)
+
+    @api.depends('card_campaign_id')
+    def _compute_card_requires_sync_count(self):
+        """Check if there's any missing or outdated card."""
+        self.card_requires_sync_count = 0
+        # no point in updating sent mailings
+        card_mailings = self.filtered(lambda mailing: mailing.card_campaign_id and mailing.state == 'draft')
+        for mailing in card_mailings:
+            TargetModel = self.env[mailing.mailing_model_real]
+            expected = TargetModel.search(self._card_parse_mailing_domain_raw())
+            actual = TargetModel.search(self._parse_mailing_domain())
+            missing_count = len(expected) - len(actual) 
+            out_of_date_count = self.env['card.card'].search_count([
+                ('campaign_id', '=', mailing.card_campaign_id.id),
+                ('res_id', 'in', expected.ids),
+                ('generation', '<', mailing.card_campaign_id.generation)
+            ])
+            mailing.card_requires_sync_count = missing_count + out_of_date_count
+
+    def action_update_cards(self):
+        """Update the cards in batches, commiting after each batch."""
+        for campaign in self.filtered(lambda mailing: mailing.state == 'draft').card_campaign_id:
+            campaign._update_cards(self._card_parse_mailing_domain_raw(), auto_commit=True)
+        return {
+            'type': 'ir.actions.act_window',
+            'res_model': 'mailing.mailing',
+            'res_id': self[0].id,
+            'view_mode': 'form',
+            'target': 'current',
+        }
+
+    def action_put_in_queue(self):
+        """Detect mismatches before scheduling."""
+        for mailing in self.filtered('card_campaign_id'):
+            if mailing.card_requires_sync_count:
+                raise exceptions.UserError(_('You should update all the cards before scheduling a mailing.'))
+        super().action_put_in_queue()
+
+    def _card_parse_mailing_domain_raw(self):
+        return super()._parse_mailing_domain()
+
+    def _parse_mailing_domain(self):
```

Comment:
That's strange ... purpose of this method is simply to parse the text version of the mailing domain. Ok we have to check recipients that still have no card linked to them, but for other processes not sure we should add this in domain ? It is something related to maketing cards, before sending, be sure cards are generated. Then it should behave like before in my opinion, aka calls to _parse_mailing_domain should not be impacted, we just want the mailing domain.

Reply:
> before sending, be sure cards are generated

Not sure if it was clear but the reason for this is that we don't want to send the mailing to recipients that don't have matching cards. As then they'll receive a broken url in their mail, probably telling them to share a card not related to them at all. This can easily happen if you schedule the mailing, and someone registers before the scheduled date (as we don't want to generate cards in the cron we can't just fill the blanks..)

I've added a new method to mailing to get the "recipient domain" and replaced calls to `_parse_mailing_domain` with it where relevant.